fdbclient/NativeAPI.actor.cpp

```diff
@@ -8013,6 +8018,75 @@ ACTOR Future<Version> setPerpetualStorageWiggle(Database cx, bool enable, LockAw
 	return version;
 }
 
+ACTOR Future<Version> checkBlobSubrange(Database db, KeyRange keyRange, Optional<Version> version) {
+	state Transaction tr(db);
+	state Version readVersionOut = invalidVersion;
+	loop {
+		try {
+			wait(success(tr.readBlobGranules(keyRange, 0, version, &readVersionOut)));
+			return readVersionOut;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+}
+
+ACTOR Future<Version> verifyBlobRangeActor(Reference<DatabaseContext> cx, KeyRange range, Optional<Version> version) {
+	state Database db(cx);
+	state Transaction tr(db);
+	state Standalone<VectorRef<KeyRangeRef>> allRanges;
+	state Key beginKey = range.begin;
+	state Key endKey = range.end;
+	state KeyRange currentChunk;
+	state Version readVersionOut = invalidVersion;
+	state int batchSize = CLIENT_KNOBS->BG_TOO_MANY_GRANULES / 2;
+	loop {
+		try {
+			wait(store(allRanges, tr.getBlobGranuleRanges(KeyRangeRef(beginKey, endKey), 20 * batchSize)));
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+
+		if (allRanges.empty()) {
+			if (beginKey < endKey) {
+				return invalidVersion;
+			}
+			return readVersionOut;
+		}
+
+		state std::vector<Future<Version>> checkParts;
+		// Chunk up to smaller ranges than this limit. Must be smaller than BG_TOO_MANY_GRANULES to not hit the limit
+		int maxChunkSize = CLIENT_KNOBS->BG_TOO_MANY_GRANULES / 2;
+		int currentChunkSize = 0;
+		for (auto& it : allRanges) {
+			if (currentChunkSize == maxChunkSize) {
+				checkParts.push_back(checkBlobSubrange(db, currentChunk, version));
```

Comment:
I might be misreading this, but I think this won't catch the case of checking [A - E):
batch 1: [A - B), [B - C)
batch 2: [D - E)

since in batch 2, currentChunk won't be empty?
We could fix this trivially by just resetting current chunk to the empty range after each batch though i think?

Reply:
Yeah you're right. I cleaned it up to make it easier to shorter and easier to understand.